src/oni2-keyboard-layout/stubs/keyboard-layout-mac.c

```diff
@@ -0,0 +1,243 @@
+#ifdef __APPLE__
+
+#import <Cocoa/Cocoa.h>
+#import <Carbon/Carbon.h>
+
+#include <caml/alloc.h>
+#include <caml/callback.h>
+#include <caml/memory.h>
+#include <caml/mlvalues.h>
+#include <caml/threads.h>
+
+#include "keyboard-layout.h"
+
+typedef struct KeycodeMapEntry {
+  uint16_t virtualKeyCode;
+  const char *dom3Code;
+} KeycodeMapEntry;
+
+#define USB_KEYMAP_DECLARATION static const KeycodeMapEntry keyCodeMap[] =
+#define USB_KEYMAP(usb, evdev, xkb, win, mac, code, id) {mac, code}
+
+#define UNWRAP_REF(ref) (Field(ref, 0))
+
+#include "keycode_converter_data.h"
+
+NSString *oni2_priv_GetCurrentKeyboardLayout() {
+  TISInputSourceRef source = TISCopyCurrentKeyboardInputSource();
+  CFStringRef sourceID = (CFStringRef)TISGetInputSourceProperty(source, kTISPropertyInputSourceID);
+
+  return (NSString *)sourceID;
+}
+
+NSString *oni2_priv_GetCurrentKeyboardLanguage() {
+  TISInputSourceRef source = TISCopyCurrentKeyboardInputSource();
+  NSArray *langs = (NSArray *)TISGetInputSourceProperty(source, kTISPropertyInputSourceLanguages);
+  NSString *lang = (NSString *)[langs objectAtIndex:0];
+
+  return lang;
+}
+
+static void notificationHandler(
+  CFNotificationCenterRef center, 
+  void *manager,
+  CFStringRef name, 
+  const void *object, 
+  CFDictionaryRef userInfo
+) {
+  // Maintain a reference to the callbacks from the OCcaml side
+  // We let OCaml manage the list since they have better abstractions than we do
+  static const value *callbackListRef = NULL;
+  // If we haven't gotten the reference yet, get it.
+  if (callbackListRef == NULL) {
+    callbackListRef = caml_named_value("oni2_KeyboardLayoutCallbackListRef");
+    // If we didn't get anything back from the OCaml runtime, return early.
+    if (callbackListRef == NULL) {
+      NSLog(@"Unable to acquire callback list reference!");
+      return;
+    }
+  }
+
+  NSString *nsLayout = oni2_priv_GetCurrentKeyboardLayout();
+  NSString *nsLanguage = oni2_priv_GetCurrentKeyboardLanguage();
+
+  value vLayout = caml_copy_string([nsLayout UTF8String]);
+  value vLanguage = caml_copy_string([nsLanguage UTF8String]);
```

Comment:
I believe these will need to be wrapped with `CAMLlocal` - I wonder how this will work with the `caml_acquire_runtime_system` below?

Reply:
We might have to change the flow a bit here so that these are properly tracked by the GC, to avoid crashes - we could have a function get the current layout, and current language, and have the notification be a simple callback that just basically says "Go ask for the latest layout and language"

Reply:
Changed in e8e0b09!